python/mxnet/gluon/contrib/estimator/estimator.py

```diff
@@ -358,10 +359,11 @@ def fit(self, train_data,
         batch_axis : int, default 0
             Batch axis to split the training data into devices.
         """
-        if not isinstance(train_data, DataLoader):
-            raise ValueError("Estimator only support input as Gluon DataLoader. Alternatively, you "
-                             "can transform your DataIter or any NDArray into Gluon DataLoader. "
-                             "Refer to gluon.data.dataloader")
+        try:
+            is_iterable = iter(train_data)
```

Comment:
Calling `iter(train_data)` can have side-effects, such as starting prefetching. See https://github.com/dmlc/gluon-nlp/blob/297e14732849f2b8025de6e2bf80ac05cab82b44/src/gluonnlp/data/stream.py#L341

Thus `iter` must only be invoked once. The enumerate call below invokes `iter` a second time implicitly. If you want to invoke `iter` manually, you need to use a while loop and `next` to iterate over the elements until `StopIteration` is raised.

Reply:
what about using `hasattr(train_data, '__iter__')` to check if the `train_data` is iterable or not?

Reply:
You need to check both. Thus `hasattr(train_data, '__iter__') or hasattr(train_data, '__getitem__')` should be fine.

See 

> An object capable of returning its members one at a time. Examples of iterables include all sequence types (such as list, str, and tuple) and some non-sequence types like dict, file objects, and objects of any classes you define with an `__iter__()` method or with a `__getitem__()` method that implements Sequence semantics.

https://docs.python.org/3/glossary.html

Reply:
ok, it makes sense to me.